### PR TITLE
feat: splitting jsonrpc using traits

### PR DESCRIPTION
**Pull request by frolvanya**
The main idea behind this PR is to address and resolve an [issue](https://docs.nearone.org/doc/readrpc-1370-compatible-release-postmortem-O2gFHCApds#h-issue-1) where certain JSON-RPC requests resulted in errors due to the removal of code that was essential for parsing incoming data. This error, specifically NearApi::Rpc::ServerError with a message indicating a "Failed to serialize JsonRpcError," occurred because Read-RPC team accidentally dropped support for legacy input parameters in our JSON-RPC implementation

Previously, our Read-RPC setup relied on a combination of structures from the near-jsonrpc and near-jsonrpc-primitives crates. However, the `Params` structure necessary for parsing various input formats was not appropriately exposed, leading to dependency issues and inadvertently removed functionality

To resolve this, I have introduced a new library in this PR, `near-jsonrpc-traits`, that provides a more modular and flexible approach to handling JSON-RPC requests. This new library separates concerns, making it easier to extend and maintain the JSON-RPC functionality without unnecessary dependencies

### CODE REVIEW

File: chain/near-jsonrpc-traits/src/handler.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
+use std::future::Future;
+
+pub trait JsonRpcHandlerExt {
+    fn send_tx(
```
Comment (khorolets) [inline chain/near-jsonrpc-traits/src/handler.rs:4]: Can you consider moving the docstrings from the `near-jsonrpc` library, at least if they were there, to this new crate? What do you think?

File: chain/near-jsonrpc-traits/src/params.rs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+use near_jsonrpc_primitives::errors::RpcParseError;
+
+use serde::de::DeserializeOwned;
+use serde_json::Value;
+
+pub trait ParamsExt<T> {
+    fn new(params: Value) -> Self;
+    fn parse(value: Value) -> Result<T, RpcParseError>
```
Comment (khorolets) [inline chain/near-jsonrpc-traits/src/params.rs:8]: I think it is better to have a default implementation here in the trait. We will not have different logic of parsing the params in both `near-jsonrpc` and ReadRPC. I also doubt anybody needs to use a different logic, but they will still be able to do so.

Do you have objections?

File: chain/jsonrpc/src/api/mod.rs [hunk 3]
```diff
@@ -109,12 +110,12 @@ mod params {
         Result<Result<T, RpcParseError>, Value>,
     );
 
-    impl<T> Params<T> {
-        pub fn new(params: Value) -> Self {
+    impl<T> ParamsExt<T> for Params<T> {
```
Comment (khorolets) [inline chain/jsonrpc/src/api/mod.rs:113]: I expected these implementations to be default ones in the `ParamsExt` trait

File: chain/jsonrpc/src/lib.rs [hunk 4]
```diff
@@ -854,7 +804,170 @@ impl JsonRpcHandler {
...
+        let congestion_info = chunk_view.header.congestion_info;
+        let congestion_level = congestion_info
+            .map(|info| info.congestion_level(config.runtime_config.congestion_control_config))
+            .unwrap_or(0.0);
+        Ok(near_jsonrpc_primitives::types::congestion::RpcCongestionLevelResponse {
+            congestion_level,
+        })
+    }
+
+    async fn changes_in_block_by_type(
```
Comment (khorolets) [inline chain/jsonrpc/src/lib.rs:870]: I believe this method should be part of the trait. It is responsible for `EXPERIMENTAL_changes` JSON RPC method

File: chain/jsonrpc/src/lib.rs [hunk 5]
```diff
@@ -854,7 +804,170 @@ impl JsonRpcHandler {
...
+            })
+            .await?;
+
+        Ok(near_jsonrpc_primitives::types::changes::RpcStateChangesInBlockResponse {
+            block_hash: block.header.hash,
+            changes,
+        })
+    }
+
+    async fn client_config(
```
Comment (khorolets) [inline chain/jsonrpc/src/lib.rs:894]: I wonder why you haven't included some public methods into a trait `JsonRpcHandlerExt`? Am I missing something? Are you missing something? :)

**Review by khorolets — COMMENTED** [review]
Great progress so far! I left a few comments to highlight some places I have concerns. I suggest we discuss a bigger picture synchronously, I will reach out to you to set a call.

File: chain/near-jsonrpc-traits/src/handler.rs (hunk 1 above)
Comment (frolvanya) [reply chain/near-jsonrpc-traits/src/handler.rs:6]: I move them, but only `validators_ordered` and `maintenance_windows` had docs

File: chain/jsonrpc/src/api/mod.rs (hunk 3 above)
Comment (frolvanya) [reply chain/jsonrpc/src/api/mod.rs]: I left trait, but moved struct&implementation from `mod.rs` to the new crate, so now everything in one file and you also can easily add own implementation using public trait

File: chain/jsonrpc/src/lib.rs (hunk 5 above)
Comment (frolvanya) [reply chain/jsonrpc/src/lib.rs]: Yep, I had a brain lag. Now I've double checked that all methods that are used in `process_basic_requests_internal` are available in trait

**Comment by frolvanya** [comment]
@race-of-sloths please, include my PR in the Race

**Comment by race-of-sloths** [comment]
@frolvanya Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
Are you going to win race-of-sloths this month? If so, you should speed up! Run, Sloth, run!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/frolvanya?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/frolvanya?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/frolvanya)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command
</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/frolvanya)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

**Comment by khorolets** [comment]
@race-of-sloths score 5

**Comment by race-of-sloths** [comment]
### 🌟 Score recorded!

@khorolets, thank you for scoring this pull request in the Race of Sloths!

**Comment by race-of-sloths** [comment]
### ❌ PR inactive
This PR has been inactive for two weeks. It's now marked as stale. 
>[!TIP]
If you think it's a mistake, you can restart the bot with `include` command to proceed 🔄

**Pull request closed without merging**